axonserver/src/main/java/io/axoniq/axonserver/message/event/EventDispatcher.java

```diff
@@ -188,68 +187,24 @@ public EventDispatcher(EventStoreLocator eventStoreLocator,
-                                                        sink.complete();
-                                                        removeTrackerInfo(trackerInfo);
-                                                    });
-
-                        trackingEventProcessors.computeIfAbsent(new ClientStreamIdentification(trackerInfo.context,
-                                                                                               trackerInfo.client),
-                                                                key -> new CopyOnWriteArrayList<>()).add(trackerInfo);
-                        logger.info("Starting tracking event processor for {}:{} - {}",
-                                    request.getClientId(),
-                                    request.getComponentName(),
-                                    request.getTrackingToken());
-                    }
-                    Sinks.Many<GetEventsRequest> eventStoreRequestFlux = eventStoreRequestFluxRef.get();
-                    if (eventStoreRequestFlux != null) {
-                        if (eventStoreRequestFlux.tryEmitNext(request).isFailure()) {
-                            eventStoreRequestFlux.tryEmitError(new RuntimeException("Unable to send the request for events."));
-                            removeTrackerInfo(eventTrackerInfoRef.get());
-                        }
-                    }
-                }, t -> {
-                    Sinks.Many<GetEventsRequest> eventStoreRequestFlux = eventStoreRequestFluxRef.get();
-                    if (eventStoreRequestFlux != null) {
-                        eventStoreRequestFlux.tryEmitError(t);
-                    }
-                    removeTrackerInfo(eventTrackerInfoRef.get());
-                }, () -> {
-                    Sinks.Many<GetEventsRequest> eventStoreRequestFlux = eventStoreRequestFluxRef.get();
-                    if (eventStoreRequestFlux != null) {
-                        eventStoreRequestFlux.tryEmitComplete();
-                    }
-                    removeTrackerInfo(eventTrackerInfoRef.get());
-                }));
+        return requestFlux.switchOnFirst((signal, rf) -> {
+            if (signal.isOnNext()) {
+                GetEventsRequest request = signal.get();
+                EventTrackerInfo trackerInfo = new EventTrackerInfo(request.getClientId(),
+                                                                    context,
+                                                                    request.getTrackingToken() - 1);
+                return eventStoreLocator.eventStore(context, request.getForceReadFromLeader())
+                                        .flatMapMany(eventStore -> eventStore.events(context,
+                                                                                     principal,
+                                                                                     rf))
+                                        .takeUntil(e -> trackerInfo.isCompleted())
```

Comment:
The takeUntil to me seems to be only executed when a new message is published on the requestFlux. If the completed field of the tracker is set to true, it is not executed, so nothing will happen.

Reply:
Correct! Stream is going to be opened until it is necessary to read a new event. This could be problematic, let me try to find an alternative.